main/src/main/scala/sbt/Defaults.scala

```diff
@@ -1956,6 +1961,11 @@ trait BuildExtra extends BuildCommon with DefExtra {
   def filterKeys(ss: Seq[Setting[_]], transitive: Boolean = false)(f: ScopedKey[_] => Boolean): Seq[Setting[_]] =
     ss filter (s => f(s.key) && (!transitive || s.dependencies.forall(f)))
 }
+object BuildExtra {
+  import scala.collection.mutable.Map
```

Comment:
This is what I'm trying to achieve at the moment. It's most likely possible, but I'm afraid this would require a lot of changes.